File: array_and_strings/two_pointers/c++/reverse_words/reverse_words.cpp
// reverse_words.cpp
#include "reverse_words.hpp"
#include <string>
#include <boost/algorithm/string.hpp>
#include <vector>

// Definition of reverseWordsLoop
void WordReverser::reverseWordsLoop(std::string &s)
{
    // Stores the index of space character we most recently reached, initialized to -1
    int mostRecentSpaceIndex = -1;
    int n = s.size();

    // Traverse the entire string 0 to n - 1
    for (int strIndex = 0; strIndex <= n; strIndex++)
    {
        // The if block runs if we have reached a space character or if we are at the end of the string 's'
        if (strIndex == n || s[strIndex] == ' ')
        {
            // Start of the word index (the very first iteration leftIndex will be -1 + 1 = 0)
            int leftIndex = mostRecentSpaceIndex + 1;
            // End of the word index, which is one before the current strIndex that points at a particular space character
            int rightIndex = strIndex - 1;

            while (leftIndex < rightIndex)
            {
                // Swap the characters and move pointers towards each other
                std::swap(s[leftIndex], s[rightIndex]);
                leftIndex++;
                rightIndex--;
            }
            // Before the current iteration ends, update 'mostRecentSpaceIndex' to the index of the current space character
            mostRecentSpaceIndex = strIndex;
        }
    }
}

// Definition of reverseWordsBoost
std::string WordReverser::reverseWordsBoost(std::string s)
{
    // Split string into a vector of words
    std::vector<std::string> vecOfWords;
    // Use token_compress_on if there's consecutive separators
    boost::split(vecOfWords, s, boost::is_any_of(" "), boost::token_compress_on);

    // Iterate through each word and reverse it
    for (auto &word : vecOfWords)
    {

        int leftIndex = 0;
        int rightIndex = word.size() - 1;

        // Loop ends when the two pointers meet in the middle
        while (leftIndex < rightIndex)
        {
            // Use std::iter_swap to swap characters
            std::iter_swap(word.begin() + leftIndex, word.begin() + rightIndex);
            leftIndex++;
            rightIndex--;
        }
    }

    // Join the vector of words into a single string
    std::string result = boost::algorithm::join(vecOfWords, " ");
    return result;
}

File: array_and_strings/two_pointers/python3/reverse_words.py
def reverse_words(s: str) -> str:
    """
    Given a string `s`, reverse the order of characters in each word within a sentence while still preserving whitespace and initial word order.

    Parameters
    ----------
    s : str
        A string containing words separated by whitespace.

    Returns
    -------
    str
        A string with the order of characters in each word reversed while preserving whitespace and initial word order.
    """
    # Split Sentence into list of words
    list_of_words = s.split(sep=" ")

    result = []
    for word in list_of_words:
        left_index = 0
        right_index = len(word) - 1

        # Split the word into list of characters
        word = list(word)

        # Loop ends when the pointer meet at middle of the array
        while left_index < right_index:
            # Effectively reversing the characters
            word[left_index], word[right_index] = word[right_index], word[left_index]
            # Move pointers towards each other
            left_index += 1
            right_index -= 1

        # Join reverse list of characters back as a single string
        word = "".join(word)
        result.append(word)

    return " ".join(result)


def main() -> int:
    s_1 = "Let's take LeetCode contest"
    s_2 = "Mr Ding"

    for s in [s_1, s_2]:
        print(f"The input string {s_1} becomes {reverse_words(s)}")

    return 0


if __name__ == "__main__":
    main()
